lib/vnet/setup.go

```diff
@@ -27,69 +28,142 @@ import (
 	"golang.zx2c4.com/wireguard/tun"
 )
 
-// Run is a blocking call to create and start Teleport VNet.
-func Run(ctx context.Context, appProvider AppProvider) error {
+// SetupAndRun creates a network stack for VNet and runs it in the background. To do this, it also
+// needs to launch an admin subcommand in the background. It returns [ProcessManager] which controls
+// the lifecycle of both background tasks.
+//
+// The caller is expected to call Close on the process manager to close the network stack, clean
+// up any resources used by it and terminate the admin subcommand.
+//
+// ctx is used to wait for setup steps that happen before SetupAndRun hands out the control to the
+// process manager. If ctx gets canceled during SetupAndRun, the process manager gets closed along
+// with its background tasks.
+func SetupAndRun(ctx context.Context, appProvider AppProvider) (*ProcessManager, error) {
 	ipv6Prefix, err := NewIPv6Prefix()
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
-
 	dnsIPv6 := ipv6WithSuffix(ipv6Prefix, []byte{2})
 
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	pm, processCtx := newProcessManager()
+	success := false
+	defer func() {
+		if !success {
+			// Closes the socket and background tasks.
+			pm.Close()
+		}
+	}()
+
+	// Create the socket that's used to receive the TUN device from the admin subcommand.
+	socket, socketPath, err := createUnixSocket()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	slog.DebugContext(ctx, "Created unix socket for admin subcommand", "socket", socketPath)
+	go func() {
+		// Keep the socket open until the process context is canceled.
+		// Closing the socket signals the admin subcommand to terminate.
+		<-processCtx.Done()
+		_ = socket.Close()
+	}()
```

Comment:
since this is tied to the process lifetime, i think we might as well make it a critical task on the process manager, and that gives us a guarantee that something will wait for the goroutine to terminate and things will fail loudly if this somehow exits early

```suggestion
	pm.AddCriticalBackgroundTask("socket closer", func() error {
		// Keep the socket open until the process context is canceled.
		// Closing the socket signals the admin subcommand to terminate.
		<-processCtx.Done()
		return trace.Wrap(socket.Close())
	})
```

Reply:
That works, but we have to do add an error aggregate:

```go
	pm.AddCriticalBackgroundTask("socket closer", func() error {
		// Keep the socket open until the process context is canceled.
		// Closing the socket signals the admin subcommand to terminate.
		<-processCtx.Done()
		return trace.NewAggregate(processCtx.Err(), socket.Close())
	})
```

This makes the task behave like other context-aware functions (as in, they return `context.Canceled` when idle when the context gets canceled). Otherwise after Ctrl+C, this would race with the other tasks – if `socket.Close()` returned with no error, the user would see `critical task "socket closer" exited prematurely`.

That's because `errgroup.Group.Wait` does not await the context by itself, it waits for the first goroutine to finish. And `processManager.Wait` lost the ability to check if the context was canceled after we moved the context outside of the struct.

I could also do something like this, I'm not sure if this is better:

```go
	pm.AddCriticalBackgroundTask("socket closer", func() error {
		// Keep the socket open until the process context is canceled.
		// Closing the socket signals the admin subcommand to terminate.
		<-processCtx.Done()
		err := processCtx.Err()
		if closeErr := socket.Close(); closeErr != nil {
			err = closeErr
		}
		return trace.Wrap(err)
	})
```

This socket closer is essentially a cleanup function, that's why I also felt like it shouldn't be deemed a critical task.

For now I'm going to turn it into a critical task and add an aggregate.

Reply:
https://github.com/golang/go/issues/23595#issuecomment-361633753

> The purpose of the `errgroup` package is to coordinate error collection with cancellation, since that's fairly subtle to get right using `sync.WaitGroup`.
If you cancel the `Context` on the first error, the subsequent errors tend not to be meaningful (they're often either `context.Canceled` or other errors that arise as a side-effect of cancellation).
> 
> If you want all of the errors (and don't want to cancel the `Context`), then you may as well just use a `sync.WaitGroup`. It's not a lot of boilerplate, and not particularly error-prone.